crates/kubelet/src/device_plugin_manager/mod.rs

```diff
@@ -0,0 +1,954 @@
+                quantity,
+                resource_name,
+                available_devices.len()
+            ));
+        }
+
+        // let endpoint = self.plugins.lock().unwrap().get(resource_name).unwrap().clone();
+        // let get_preferred_allocation_available = match &endpoint.register_request.options {
+        //     None => false,
+        //     Some(options) => options.get_preferred_allocation_available,
+        // };
+
+        // if get_preferred_allocation_available {
+        //
+        // }
+
+        // TODO: support preferred allocation
+        // For now, reserve first N devices where N = quantity by adding them to allocated map
+        let devices_to_allocate: Vec<String> = available_devices[..quantity].to_vec();
+        // ??: is there a cleaner way to map `allocated_devices.insert(dev.clone())` from bool to
+        // () so can remove {} block:
+        devices_to_allocate.iter().for_each(|dev| {
+            allocated_devices.insert(dev.clone());
+        });
+        allocated_devices_map.insert(resource_name.to_string(), allocated_devices);
+
+        Ok(devices_to_allocate)
+    }
+}
+
+/// Returns the device IDs of all healthy devices that have yet to be allocated.
+fn get_available_devices(
+    devices: &EndpointDevicesMap,
+    allocated_devices: &HashSet<String>,
+) -> Vec<String> {
+    let healthy_devices = devices
+        .iter()
+        .filter_map(|(dev_id, dev)| {
+            if dev.health == HEALTHY {
+                Some(dev_id.clone())
+            } else {
+                None
+            }
+        })
+        .collect::<HashSet<String>>();
+    healthy_devices
+        .difference(allocated_devices)
+        .into_iter()
+        .cloned()
+        .collect::<Vec<String>>()
+}
+
+/// Hosts the device plugin `Registration` service (defined in the device plugin API) for a `DeviceManager`.
+/// Upon device plugin registration, reaches out to its `DeviceManager` to validate the device plugin
+/// and establish a connection with it.
+#[derive(Clone)]
+pub struct DeviceRegistry {
+    device_manager: Arc<DeviceManager>,
+}
+
+impl DeviceRegistry {
+    /// Returns a new `DeviceRegistry` with a reference to a `DeviceManager` to handle each registered device plugin.
+    pub fn new(device_manager: Arc<DeviceManager>) -> Self {
+        DeviceRegistry { device_manager }
+    }
+}
+#[async_trait::async_trait]
+
+impl Registration for DeviceRegistry {
+    async fn register(
+        &self,
+        request: tonic::Request<RegisterRequest>,
+    ) -> Result<tonic::Response<Empty>, tonic::Status> {
+        let register_request = request.get_ref();
+        // Validate
+        self.device_manager
+            .validate(register_request)
+            .await
+            .map_err(|e| tonic::Status::new(tonic::Code::InvalidArgument, format!("{}", e)))?;
+        // Create a list and watch connection with the device plugin
+        // TODO: should the manager keep track of threads?
```

Comment:
Which threads are you worried about here?

Reply:
Each time a device plugin registers with the `DeviceRegistry` `register` service, `create_endpoint` is called which spawns a thread to create a ListAndWatch connection with the device plugin. This connection lasts as long as the device plugin exists, which in the case of hardware resources should be forever. I am wondering if `create_endpoint` should return a handle to that thread and all handles could be stored in the `DeviceRegistry`.